# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [33]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [34]:
os.getcwd()
os.chdir(r'C:\Users\2bogu\Desktop\Sringboard_Materials\guided capstone mat\data')
df = pd.read_csv('step3_output.csv', index_col=0)
df.head()

Name       state  summit_elev  vertical_drop  \
1            Eaglecrest Ski Area      Alaska       2600.0         1540.0   
2               Hilltop Ski Area      Alaska       2090.0          294.0   
4            Sunrise Park Resort     Arizona      11100.0         1800.0   
5  Yosemite Ski & Snowboard Area  California       7800.0          600.0   
6                  Bear Mountain  California       8805.0         1665.0   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
1    0.0        0.0        0.0        0.0   0.0     0.0  ...   
2    0.0        0.0        0.0        0.0   0.0     1.0  ...   
4    0.0        0.0        0.0        1.0   2.0     3.0  ...   
5    0.0        0.0        0.0        0.0   0.0     1.0  ...   
6    0.0        0.0        0.0        2.0   1.0     2.0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
4              800.0            80.0             115.0       49.0   
5               88.0             0.0             110.0       84.0   
6              198.0           198.0             122.0       76.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
1            350.0     47.000000      53.00000               90.0   
2             69.0     30.000000      34.00000              152.0   
4            250.0     74.000000      78.00000              104.0   
5            300.0     47.000000      47.00000              107.0   
6            100.0     57.916957      64.16681              130.0   

   NightSkiing_ac  clusters  
1             0.0         0  
2            30.0         0  
4            80.0         2  
5             0.0         1  
6             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [36]:
dfo = df[['state']]
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [38]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [40]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [41]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [42]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [43]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [44]:
import sklearn.metrics
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

-3.3312268851501085e+20 31019340171.160862


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [45]:
lm.intercept_

-7783879308.041758

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [46]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfcoe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfcoe.sort_values(ascending=False, by=['Coefficient']).head(10)

Coefficient
fastSixes       8.762081e+12
fastEight       6.185674e+12
trams           5.130946e+12
total_chairs    4.612489e+12
double          2.436670e+12
surface         2.175887e+12
triple          2.003859e+12
quad            1.418499e+12
fastQuads       1.368791e+12
state_Michigan  6.265902e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [47]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df[df.columns.drop(list(df.filter(regex='state')))]
X = X.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [48]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [49]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [50]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

In [51]:
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

0.6959441471403798 6.348645827530874


In [52]:
lm.intercept_

59.517785179741

In [53]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfcoe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfcoe.sort_values(ascending=False, by=['Coefficient']).head(10)

Coefficient
AdultWeekday         13.173677
SkiableTerrain_ac     2.840798
clusters              2.382784
vertical_drop         2.350428
quad                  1.956536
daysOpenLastYear      1.508513
Runs                  1.291309
surface               1.214620
projectedDaysOpen     1.204303
triple                1.058218

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [54]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df[df.columns.drop(list(df.filter(regex='state')))]
X = X.drop(['Name','AdultWeekend','summit_elev'], axis=1) #AND base_elev if it was there

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [55]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [56]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [57]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

In [58]:
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

0.6964701630020269 6.333027236718


In [59]:
lm.intercept_

59.515329135526926

In [60]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfcoe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfcoe.sort_values(ascending=False, by=['Coefficient']).head(10)

Coefficient
AdultWeekday         13.174591
SkiableTerrain_ac     2.887166
vertical_drop         2.442870
clusters              1.980375
quad                  1.955015
daysOpenLastYear      1.514394
Runs                  1.309548
surface               1.220477
projectedDaysOpen     1.213964
triple                1.082437

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -3.3312268851501085e+20|31019340171.160862 |-|
| Model 2. | 0.6959441471403798|6.348645827530874 |'state'|
| Model 3. | 0.6964701630020269|6.333027236718 |'state','summit_elev','base_elev'|

Model Selection: Model 3